### PR TITLE
Run tests for all submissions for a single assignment on demand

### LINKED ISSUE #430: Running tests for all students on-demand
Currently, QuickFeed won’t run the tests for the students at the due date automatically.

We can manually postpone publishing the tests until after the due date, which would prevent students from seeing their scores before the due date of an assignment.

I can think of three alternative solutions:
1. Add a button in the teacher panel to allow a teacher to trigger test execution for all students
2. Trigger test execution for all students when the `tests` repo is updated with a new assignment
3. Implement a "timeout" mechanism to run the tests at a given deadline

We already have a `Rebuild` button that operates per student. Hence, I think it would be easy to add a single button and an RPC call to `Rebuild All` student tests.

The second alternative should also be straightforward to implement.

The third alternative will require more work. Not clear how much.

One issue, though. If the idea is to release the score at a due date but still run tests and let students see output -- to help debug things, then the third alternative is perhaps the way to go. Otherwise, it could be achieved with the first two approaches and manually publishing tests that do not generate score objects but generate `ResultInfo` to share with the users. At the due date, publish tests that generate score objects.

### PR DESCRIPTION
**Pull request by 0xf8f8ff**
Resolves #430.

Running tests for all submissions from the Assignment view:

image

### CODE REVIEW

File: web/autograder_service.go [hunk 1]
```diff
@@ -581,13 +581,31 @@ func (s *AutograderService) RebuildSubmission(ctx context.Context, in *pb.Rebuil
...
 	}
-	submission, err := s.rebuildSubmission(ctx, in)
+	submission, err := s.rebuildSubmission(in)
 	if err != nil {
 		return nil, err
 	}
 	return submission, nil
 }
 
+// RebuildSubmissions runs tests for all submissions for the given assignment ID.
```
Comment (meling) [inline web/autograder_service.go:591]: Add access policy to the doc. I guess only teachers and admin are allowed to call this.

File: web/autograder_service.go [hunk 2]
```diff
@@ -581,13 +581,31 @@ func (s *AutograderService) RebuildSubmission(ctx context.Context, in *pb.Rebuil
...
+func (s *AutograderService) RebuildSubmissions(ctx context.Context, in *pb.AssignmentRequest) (*pb.Void, error) {
```
Comment (meling) [inline web/autograder_service.go:592]: Can we add some tests for this?

File: web/rebuild.go [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
 
 	pb "github.com/autograde/quickfeed/ag"
 	"github.com/autograde/quickfeed/ci"
 	"github.com/gosimple/slug"
+	"golang.org/x/sync/errgroup"
 )
 
 // rebuildSubmission rebuilds the given assignment and submission.
-func (s *AutograderService) rebuildSubmission(ctx context.Context, request *pb.RebuildRequest) (*pb.Submission, error) {
+func (s *AutograderService) rebuildSubmission(request *pb.RebuildRequest) (*pb.Submission, error) {
```
Comment (meling) [inline web/rebuild.go:14]: Can we add some tests for this?

File: web/rebuild.go [hunk 4]
```diff
@@ -46,6 +48,29 @@ func (s *AutograderService) rebuildSubmission(ctx context.Context, request *pb.R
 	return s.db.GetSubmission(&pb.Submission{ID: request.GetSubmissionID()})
 }
 
+func (s *AutograderService) rebuildSubmissions(request *pb.AssignmentRequest) error {
```
Comment (meling) [inline web/rebuild.go:51]: Can we add some tests for this?

File: web/rebuild.go [hunk 5]
```diff
@@ -46,6 +48,29 @@ func (s *AutograderService) rebuildSubmission(ctx context.Context, request *pb.R
...
+	log.Printf("Running tests for all submissions for assignment ID %d of course ID %d\n", request.AssignmentID, request.CourseID)
```
Comment (meling) [inline web/rebuild.go:52]: Use `s.logger.Debugf`.

File: web/rebuild.go [hunk 6]
```diff
@@ -46,6 +48,29 @@ func (s *AutograderService) rebuildSubmission(ctx context.Context, request *pb.R
...
+	for _, submission := range submissions {
+		rebuildRequest.SubmissionID = submission.ID
+		errgrp.Go(func() error {
+			_, err := s.rebuildSubmission(rebuildRequest)
+			return err
+		})
+	}
+	err = errgrp.Wait()
+	total := time.Since(start)
+	log.Println("Finished running all tests, took ", total)
```
Comment (meling) [inline web/rebuild.go:70]: Use `s.logger.Debugf`.

**Review by meling — CHANGES_REQUESTED** [review]

**Comment by meling** [comment]
We also need the rebuild button for the regular results... Not just the manual grading part... (referring to the picture). Do we need to rebuild for manual grading??

**Comment by 0xf8f8ff** [comment]
> We also need the rebuild button for the regular results... Not just the manual grading part... (referring to the picture). Do we need to rebuild for manual grading??

This button only shows for regular labs, manually graded labs do not have any tests and don't need rebuilds. For comparison: (lab2 is a regular lab with automated tests, lab3 is for manual grading):
image

File: web/autograder_service.go [hunk 7]
```diff
@@ -581,13 +581,31 @@ func (s *AutograderService) RebuildSubmission(ctx context.Context, in *pb.Rebuil
 		s.logger.Errorf("ApproveSubmission failed: submitter has no access to the course")
 		return nil, status.Error(codes.PermissionDenied, "submitter has no course access")
 	}
-	submission, err := s.rebuildSubmission(ctx, in)
+	submission, err := s.rebuildSubmission(in)
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply web/autograder_service.go:584]: I see that we actually log such errors in all the other methods in `autograder_service.go` and logging is only missing in these two. Wouldn't it be better to add logging here?

**Comment by meling** [comment]
> > We also need the rebuild button for the regular results... Not just the manual grading part... (referring to the picture). Do we need to rebuild for manual grading??
> 
> This button only shows for regular labs, manually graded labs do not have any tests and don't need rebuilds. For comparison: (lab2 is a regular lab with automated tests, lab3 is for manual grading):
> image

Ok. But should this button also be shown when selecting a lab in the left-hand menu bar??

File: web/autograder_service.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (meling) [reply web/autograder_service.go:584]: Yes, agree. On the other hand, we should probably add a logging-interceptor that logs all gRPC method calls and errors. I will create a separate issue for that. So go ahead and add logging here for now.

**Comment by 0xf8f8ff** [comment]
> > > We also need the rebuild button for the regular results... Not just the manual grading part... (referring to the picture). Do we need to rebuild for manual grading??
> > 
> > 
> > This button only shows for regular labs, manually graded labs do not have any tests and don't need rebuilds. For comparison: (lab2 is a regular lab with automated tests, lab3 is for manual grading):
> > image
> 
> Ok. But should this button also be shown when selecting a lab in the left-hand menu bar??

Hm, where exactly do you think it should be placed? Here was the only place I could think of where it is easy to see which assignment it refers to.

File: web/autograder_service.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply web/autograder_service.go:585]: Added.

File: web/autograder_service.go (hunk 1 above)
Comment (0xf8f8ff) [reply web/autograder_service.go:593]: Fixed.

File: web/autograder_service.go (hunk 2 above)
Comment (0xf8f8ff) [reply web/autograder_service.go:595]: Added for all three.

File: web/rebuild.go (hunk 5 above)
Comment (0xf8f8ff) [reply web/rebuild.go]: Fixed.

**Review by meling — APPROVED** [review]

**Pull request merged**